keps/sig-node/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,973 @@
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP aims to add a new field named `maxRestartTimesOnFailure` to podSpec which helps to control
+the restart times of a Pod when Pod's `restartPolicy` is set to `OnFailure`.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, we have `backoffLimit` to fail a Job after a certain amount of retries, which helps to
+prevent unnecessary restarts. `Pod` is also eager for this feature to avoid infinite restarting
+due to a system internal error etc. What's more, considering `Job` or other workloads are working
+at a higher level, adding `maxRestartTimesOnFailure` to `Pod` might be reusable.
+
+As a matter of fact, it's a long-standing issue and strongly needed by the community, see [issue](https://github.com/kubernetes/kubernetes/issues/65797).
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add a new field `maxRestartTimesOnFailure` to `pod.spec` which only applies when `restartPolicy` is set to `OnFailure`
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+No.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new field `maxRestartTimesOnFailure` to the `pod.spec`. `maxRestartTimesOnFailure` only applies
```

Comment:
How confident are we that we won't later want a "per-unit-time" aspect (e.g."10 failures per hour") ?

Reply:
But this is something I worried about like we have a pod running for a really long time, although it's a well-written application, it can still be restart for several times. Then we'll mark it failed then? Add it as another risk.

Reply:
> If it applies to Always, then it sounds like we'll always restart the pod until we met the maxRestartTimes, then `restartPolicy=Always` has no difference with `restartPolicy=OnFailure`.

I think it's a clear semantic:

* "Always" means that containers will be restarted, _regardless of the exit status_, until they hit the hard-stop criteria, such as `maxRestarts`.
* "OnFailure" means that containers will be restarted _if and only if the exit staus indicates failure_, until they hit the hard-stop criteria, such as `maxRestarts`.
* "Never" means that containers will not be restarted for any reason.

Am I missing something?

> That's a different perspective, I haven't thought about that. But how to define the slot, because we don't record when but how many times the container fails right now.

If we wanted a max restart rate, we would have to do a lot more tracking work.  I am not at all convinced it's useful, so maybe this is "considered and discarded" ? :)

I don't understand the last point you made - if someone sets the `maxRestarts` field, they seem to know what they are doing, no?

Reply:
+1 to Tim - I think that generalizing it to "Always" is natural and instead of making the API narrow, let's make it more generic.

Reply:
> But why we want to restart a completed pod(exit status is 0)

In the case of `restartPolicy=Always`, that is literally what the user asked for.  Maybe we never should have added that, and only had "OnFailure" (just don't `exit(0)`), but we DO have it, so ...

> Maybe we can start with maxRestarts rather than maxRestartsOnFailure in the field name(but only applies when restartPolicy=OnFailure currently), in case we want to extend in the future.

I could live with it in Alpha, but I think it needs to be complete to move past that, and I'd RATHER see it complete from the start.  I'm not intimately familiar with the restarting code - if it turns out to be SIGNIFICANTLY more complicated to do this "right" (general), then can we talk about that?  

>  I believe users want to configure the field to avoid unrecoverable restarts, but sometimes we may hit the max restart times just because the workloads run for a really long time

I agree with this, but have no evidence.  If this is true, we could add a coarse form of history (e.g. clear the counter every 24 hours, or track reasonable-N number of restarts with timestamps).

To really REALLY do this, we would probably want to categorize different failures (count local OOM kills, system OOM kills, signals, and exit codes in different accumulators and allow different policies).  I can imagine that API, but we don't want to speculate, so we should wait until we have real use cases.

I think we start with as little as we can get away with, one field in spec, and see how that holds up - unless we have more concrete information.